File: repurpose/gen-images.js
// gen-images.js — CANONICAL pool-managed ChatGPT image generator. Supersedes gen-batch.js (hardcoded
// persistent chat) and gen-batch-freshchat.js (always fresh). Consults ../chatgpt-image-chats.json via
// chat-pool.js: reuse the active chat for the purpose while count<cap, else open a fresh chatgpt.com/
// chat and register it; rotate automatically at the cap; never share a chat across purposes.
//
// Usage: node gen-images.js --list=<items.json> --prefix=x-tweets|yt-posts|ig-single|ig-carousel
//   items.json: [{ "image_id":"ab12cd34", "slug":"my-slug", "prompt":"...", "ref":"C:\\...png"(optional) }]
//   out: schedule-tweets/images/<x|yt|ig>/<prefix>-<image_id>-<slug>.png   (skips existing)
const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');
const pool = require('./chat-pool');

const PROFILE_DIR = 'C:\\Users\\mnede\\AppData\\Local\\Google\\Chrome\\chatgpt-profile';
const IMG_BASE = 'C:\\Users\\mnede\\Documents\\Claude\\social-media\\schedule-tweets\\images';
const IMAGE_URL_PATTERN = 'estuary/content';
const SEL = { composer: '#prompt-textarea, div[contenteditable="true"][data-id]' };

function args() { const a = {}; for (const x of process.argv.slice(2)) { const m = x.match(/^--([^=]+)=(.*)$/); if (m) a[m[1]] = m[2]; } return a; }
const A = args();
const LIST = JSON.parse(fs.readFileSync(A.list, 'utf-8'));
const PREFIX = A.prefix || 'x-tweets';
const PURPOSE = PREFIX; // purpose == prefix (x-tweets, yt-posts, ig-single, ig-carousel)
const SUBDIR = PREFIX === 'x-tweets' ? 'x' : PREFIX === 'yt-posts' ? 'yt' : (PREFIX === 'ig-carousel' || PREFIX === 'ig-single') ? 'ig' : 'x';
const OUTDIR = path.join(IMG_BASE, SUBDIR);

async function composerLoaded(page, ms = 12000) {
  try { await page.locator(SEL.composer).first().waitFor({ timeout: ms }); return true; }
  catch { return false; }
}
async function uploadRef(page, ref) {
  try {
    const fi = page.locator('input[type="file"]').first();
    await fi.waitFor({ state: 'attached', timeout: 8000 });
    await fi.setInputFiles(ref, { timeout: 8000 });
    await page.waitForTimeout(4000); return true;
  } catch (e) { console.log('   ref upload failed:', e.message.split('\n')[0]); return false; }
}

// ---- RELOAD-based capture (ported from generate-broll-reload.js, 2026-07-11) ----
// The live streaming DOM in an automation-detected browser sometimes NEVER surfaces the finished
// image (it just spins), but a PAGE RELOAD reveals it. So per image we SEND ONCE, poll the live DOM
// up to 80s, and if still nothing, RELOAD to capture. We NEVER re-send (a re-send = duplicate gen).
// Capture keys on the STABLE estuary file_id (id=file_...), which survives reloads, so we always grab
// the NEW image (the file_id not present before we sent) regardless of DOM order / lazy-load.
const getGenImgs = (page) => page.evaluate(() =>
  Array.from(document.querySelectorAll('img')).map(i => i.src)
    .filter(s => s.includes('estuary/content') || s.includes('oaiusercontent')));
const fileId = (src) => { const m = src.match(/id=(file_[A-Za-z0-9]+)/); return m ? m[1] : src; };

async function gotoChat(page, url) {
  await page.goto(url);
  await page.waitForLoadState('domcontentloaded');
  await composerLoaded(page, 30000);
  await page.waitForTimeout(3000);
  try { await page.evaluate(() => window.scrollTo(0, document.body.scrollHeight)); } catch {}
  await page.waitForTimeout(500);
}
// ChatGPT occasionally throws a full-screen modal ("Compare responses" A/B eval) that overlays the
// composer and hangs the next prompt. Dismiss it: Escape, then a Skip/close button, then reload.
async function dismissDialog(page) {
  try {
    const open = async () => (await page.locator('div[role="dialog"][data-state="open"], div[role="dialog"]').count()) > 0;
    if (!(await open())) return false;
    console.log('   modal dialog present -> dismissing');
    for (let i = 0; i < 3; i++) { await page.keyboard.press('Escape'); await page.waitForTimeout(700); if (!(await open())) return true; }
    for (const re of [/skip/i, /no thanks/i, /close/i, /dismiss/i, /done/i, /prefer/i]) {
      const b = page.locator('div[role="dialog"] button', { hasText: re }).first();
      if (await b.count()) { await b.click({ timeout: 4000 }).catch(() => {}); await page.waitForTimeout(700); if (!(await open())) return true; }
    }
    const b0 = page.locator('div[role="dialog"] button').first();
    if (await b0.count()) { await b0.click({ timeout: 4000 }).catch(() => {}); await page.waitForTimeout(700); }
    if (await open()) { await page.reload().catch(() => {}); await page.waitForLoadState('domcontentloaded').catch(() => {}); await page.waitForTimeout(3000); }
    return !(await open());
  } catch { return false; }
}

async function genOne(page, item) {
  const outPath = path.join(OUTDIR, `${PREFIX}-${item.image_id}-${item.slug}.png`);
  if (fs.existsSync(outPath)) { console.log(`SKIP (exists) ${item.slug}`); return 'skip'; }
  await dismissDialog(page);
  const composer = page.locator(SEL.composer).first();
  await composer.click(); await page.waitForTimeout(600);
  // Upload the --reference FIRST, THEN snapshot. ORDER IS LOAD-BEARING (fixed 2026-07-14): the
  // uploaded ref must be part of the `before` baseline, or it looks like a brand-new file_id and
  // pickNew captures OUR OWN UPLOAD as the result. That bug shipped a byte-identical copy of
  // kaspa-logo.png as a finished tweet image, and cascaded (each later item then grabbed the
  // previous item's late render, shifting filenames by one). The old estuary-preference guard
  // below is NOT sufficient on its own: it assumed an uploaded ref is served from oaiusercontent,
  // but ChatGPT serves uploads from estuary/content too, so the ref slipped straight through it.
  if (item.ref) await uploadRef(page, item.ref);
  // snapshot the file_ids already in the chat (incl. any just-uploaded ref) so the new image is the unseen file_id
  const before = new Set((await getGenImgs(page)).map(fileId));
  await composer.click();
  for (const ch of item.prompt) { await page.keyboard.type(ch); await page.waitForTimeout(Math.floor(Math.random() * 25) + 45); }
  await page.waitForTimeout(Math.floor(Math.random() * 4000) + 6000);
  await page.keyboard.press('Enter');
  // A fresh chat navigates chatgpt.com/ -> /c/<id> shortly AFTER the first send; capture that url so a
  // reload targets the generating chat (reloading chatgpt.com/ would abandon it).
  let convUrl = page.url();
  for (let i = 0; i < 20 && !/\/c\//.test(convUrl); i++) { await page.waitForTimeout(1000); convUrl = page.url(); }

  // POST-SEND RE-BASELINE (fixed 2026-07-14). THE root fix for capturing our own uploaded --reference.
  // Uploading the ref before the first snapshot is NOT enough: the attachment is not in the DOM as an
  // estuary <img> until the message is actually POSTED, so it re-appears as an "unseen" file_id right
  // after send and pickNew grabs it. It is a RACE (whoever lands first), which is why it looked
  // intermittent: 2 of 3 slides rendered fine and the 3rd came back as the exemplar.
  // A generation NEVER completes in ~3s, so anything on the page 3s after send is by definition NOT
  // our result: fold it all into `before` and it can never be picked again.
  // NOTE: a byte/md5 comparison against the ref file does NOT catch this - the CDN re-encodes the PNG,
  // so bytes/md5 differ while the pixels are identical. That false-negative is exactly why the
  // 2026-07-11 session wrongly concluded "ChatGPT reproduces the exemplar verbatim" and logged it as an
  // unresolved model quirk; it was this capture bug all along.
  await page.waitForTimeout(3000);
  for (const s of await getGenImgs(page)) before.add(fileId(s));

  const pickNew = async () => {
    const imgs = await getGenImgs(page);
    const news = imgs.filter(s => !before.has(fileId(s)));
    // Prefer estuary/content (the GENERATED image). oaiusercontent ALSO matches an UPLOADED
    // --reference image, so without this preference pickNew can grab the uploaded exemplar/logo
    // instead of the result (it did: carousel slides captured the ref exemplars verbatim). Only
    // fall back to a non-estuary url when there is no estuary image at all.
    const est = news.filter(s => s.includes('estuary/content'));
    const cand = est.length ? est : news;
    return cand.length ? cand[cand.length - 1] : null;   // newest unseen generated image
  };
  const finish = async (src) => {
    let buf = null;
    try { const r = await page.request.get(src); if (r.ok()) buf = await r.body(); } catch {}
    if (!buf || buf.length < 5000) return false;
    // MECHANICAL GATE: never accept our own uploaded --reference as the generated output. An image
    // model cannot reproduce an input byte-for-byte, so identical bytes prove a mis-capture, not a
    // render. Rejecting here makes pickNew keep polling for the real image instead of shipping the ref.
    if (item.ref && fs.existsSync(item.ref)) {
      const rb = fs.readFileSync(item.ref);
      if (rb.length === buf.length && rb.equals(buf)) {
        console.log(`   REJECT: captured the uploaded reference, not a render (${item.slug}) - still waiting`);
        return false;
      }
    }
    for (const sib of fs.readdirSync(OUTDIR).filter(f => f.endsWith('.png'))) {
      const sb = fs.readFileSync(path.join(OUTDIR, sib));
      if (sb.length === buf.length && sb.equals(buf)) { console.log(`FAIL (dup of ${sib}) ${item.slug}`); return false; }
    }
    fs.writeFileSync(outPath, buf);
    console.log(`OK ${item.slug} (${(buf.length / 1024).toFixed(0)} KB)`);
    return true;
  };

  // PHASE 1: poll the LIVE DOM up to 80s (fast path — image usually appears naturally).
  const t0 = Date.now();
  while (Date.now() - t0 < 80000) {
    await page.waitForTimeout(5000);
    let src = await pickNew();
    if (src) { await page.waitForTimeout(3000); src = (await pickNew()) || src; if (await finish(src)) { await page.waitForTimeout(2000); return true; } }
  }
  // PHASE 2: hung past 80s -> RELOAD to surface the (server-side-finished) image.
  console.log(`   >80s no new image in live DOM (hung) -> reloading to capture ${item.slug}`);
  const reloadUrl = /\/c\//.test(convUrl) ? convUrl : page.url();
  const MAX = 4 * 60 * 1000, t1 = Date.now();
  while (Date.now() - t1 < MAX) {
    await gotoChat(page, reloadUrl);
    await dismissDialog(page);
    const src = await pickNew();
    if (src && await finish(src)) { await page.waitForTimeout(2000); return true; }
    await page.waitForTimeout(15000);
  }
  console.log(`FAIL (timeout) ${item.slug}`);
  return false;
}

(async () => {
  fs.mkdirSync(OUTDIR, { recursive: true });
  console.log(`gen-images: ${LIST.length} images | purpose="${PURPOSE}" | cap=${pool.cap()} | current count=${pool.countFor(PURPOSE)}`);
  const browser = await chromium.launchPersistentContext(PROFILE_DIR, {
    channel: 'chrome', headless: false, ignoreDefaultArgs: ['--enable-automation'],
    args: ['--disable-blink-features=AutomationControlled'], viewport: null });
  await browser.addInitScript(() => Object.defineProperty(navigator, 'webdriver', { get: () => undefined }));
  const page = await browser.newPage();

  let navigatedUrl = null;   // the /c/ chat we're currently on (null = sitting on a fresh chatgpt.com/)
  let pendingFresh = false;  // opened a fresh chat; register its /c/ url after the first successful gen
  const route = `**/*${IMAGE_URL_PATTERN}*`;

  async function openFresh() {
    await page.route(route, r => r.abort());
    await page.goto('https://chatgpt.com/');
    await page.waitForLoadState('domcontentloaded');
    await composerLoaded(page, 30000);
    await page.waitForTimeout(2500);
    await page.unroute(route);
    navigatedUrl = null; pendingFresh = true;
    console.log('  opened a FRESH chat (will register its URL after first image)');
  }
  async function ensureChat() {
    const active = pool.getActiveUrl(PURPOSE);
    if (!active) { if (!(navigatedUrl === null && pendingFresh)) await openFresh(); return; }
    if (active === navigatedUrl) return;            // already on it, has room
    // navigate to the active pool chat (block its history images during load)
    await page.route(route, r => r.abort());
    await page.goto(active);
    await page.waitForLoadState('domcontentloaded');
    const ok = await composerLoaded(page, 12000);
    await page.waitForTimeout(2500);
    await page.unroute(route);
    if (!ok) { console.log('  stored chat unreachable/deleted -> markDead + fresh'); pool.markDead(PURPOSE); await openFresh(); return; }
    navigatedUrl = active; pendingFresh = false;
    console.log(`  reusing pool chat (${pool.countFor(PURPOSE)}/${pool.cap()}): ${active}`);
  }

  let ok = 0;
  for (const item of LIST) {
    const outPath = path.join(OUTDIR, `${PREFIX}-${item.image_id}-${item.slug}.png`);
    if (fs.existsSync(outPath)) { console.log(`SKIP (exists) ${item.slug}`); ok++; continue; }
    await ensureChat();
    let r = await genOne(page, item);
    if (r === false) { console.log('   retry once...'); r = await genOne(page, item); }
    if (r === true) {
      if (pendingFresh) {
        // API-confirmed registration + gated rename ("b-roll:"/"social:" title) — never trust
        // page.url() alone, its id can diverge from the real conversation id (2026-07-22).
        const reg = await pool.confirmAndRegister(page, PURPOSE);
        if (reg) { navigatedUrl = reg.url; pendingFresh = false; }
      }
      pool.recordImage(PURPOSE);
      ok++;
    }
  }
  console.log(`\nDone: ${ok}/${LIST.length} | ${PURPOSE} chat now ${pool.countFor(PURPOSE)}/${pool.cap()}`);
  // Delete rotated-out/dead chats while the browser is still open (registry `retired` list).
  // A sweep failure never blocks the run — the chat stays queued for the next sweep.
  try { await require('./chat-delete').sweepRetired(page); }
  catch (e) { console.log('  [chat-delete] sweep error: ' + e.message.split('\n')[0]); }
  await browser.close();
})().catch(e => { console.error(e); process.exit(1); });

File: repurpose/gen_images.py
# gen_images.py — CANONICAL pool-managed ChatGPT image generator (Python port of
# gen-images.js, Wave 6 Lane 3 migration 2026-08-09; the JS twin is FROZEN rollback).
#
# Drives the SHARED chatgpt-profile Chrome via Playwright. Callers that can overlap a
# remotion-builder MUST hold the `chatgpt` stage lock around the run (lane3_batch.py
# does; the lock stays at the caller like it did for the JS).
#
# Usage: python gen_images.py --list <items.json> --prefix x-tweets|yt-posts|ig-single|ig-carousel
#   items.json: [{ "image_id":"ab12cd34", "slug":"my-slug", "prompt":"...",
#                  "ref": "C:\\...png" | ["...png", ...] (optional) }]
#   out: <images-base>/<x|yt|ig>/<prefix>-<image_id>-<slug>.png   (skips existing)
#
# Machine lines for the graph (per item + final):
#   IMG OK purpose=<p> id=<id> slug=<slug> bytes=<n> out=<path>
#   IMG SKIP purpose=<p> id=<id> slug=<slug> (exists)
#   IMG FAIL purpose=<p> id=<id> slug=<slug> reason=<...>
#   PROGRESS <n>%  ·  GEN DONE ok=N skip=N fail=N
# Exit 1 if any item failed (the graph HALTS; zero-retry doctrine lives above us —
# within a run each item gets the JS twin's single in-run retry, which is
# side-effect-safe for image gen, unlike posters).
#
# Ported hardening (2026-07-11 -> 2026-07-30, byte-faithful in behavior):
#   - RELOAD-based capture: send ONCE, poll the live DOM up to 80s, then RELOAD to
#     surface the server-side-finished image (automation-detected DOM can spin forever).
#     NEVER re-send on a hung DOM (a re-send = duplicate generation).
#   - Capture keys on the STABLE estuary file_id (id=file_...), which survives reloads.
#   - Ref upload BEFORE the baseline snapshot + POST-SEND RE-BASELINE at +3s: an
#     uploaded reference re-appears as an "unseen" file_id right after send and a naive
#     pick captures OUR OWN UPLOAD as the result (shipped a byte-identical kaspa-logo
#     as a finished image once, and cascaded filenames by one).
#   - Prefer estuary/content urls (generated) over oaiusercontent (uploads).
#   - MECHANICAL GATE: reject a capture byte-identical to any uploaded ref, and any
#     byte-dup of an existing sibling PNG (every image is unique).
#   - Modal dismissal for the full-screen "Compare responses" A/B overlay.
#   - Fresh-chat registration via chat_pool.confirm_and_register (API-confirmed id,
#     gated rename) after the FIRST successful image; count via record_image.
#   - End of run: chat_delete.sweep_retired (a sweep failure never blocks the run).

import argparse
import json
import random
import re
import sys
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
import chat_pool as pool  # noqa: E402
import chat_delete  # noqa: E402

sys.stdout.reconfigure(encoding="utf-8", errors="replace")
sys.stderr.reconfigure(encoding="utf-8", errors="replace")

PROFILE_DIR = r"C:\Users\mnede\AppData\Local\Google\Chrome\chatgpt-profile"
REPO_ROOT = Path(__file__).resolve().parents[1]
IMG_BASE_DEFAULT = REPO_ROOT / "schedule-tweets" / "images"
IMAGE_URL_PATTERN = "estuary/content"
COMPOSER_SEL = '#prompt-textarea, div[contenteditable="true"][data-id]'
FILE_ID_RE = re.compile(r"id=(file_[A-Za-z0-9]+)")

_GET_GEN_IMGS_JS = """
() => Array.from(document.querySelectorAll('img')).map(i => i.src)
  .filter(s => s.includes('estuary/content') || s.includes('oaiusercontent'))
"""


def subdir_for(prefix: str) -> str:
    return {"x-tweets": "x", "yt-posts": "yt",
            "ig-carousel": "ig", "ig-single": "ig"}.get(prefix, "x")


def ref_list(ref):
    """`ref` may be one path or a list (multi-coin lineups upload several)."""
    if not ref:
        return []
    return list(ref) if isinstance(ref, (list, tuple)) else [ref]


def file_id(src: str) -> str:
    m = FILE_ID_RE.search(src)
    return m.group(1) if m else src


def get_gen_imgs(page):
    return page.evaluate(_GET_GEN_IMGS_JS)


def composer_loaded(page, ms=12000) -> bool:
    try:
        page.locator(COMPOSER_SEL).first.wait_for(timeout=ms)
        return True
    except Exception:
        return False


def upload_ref(page, ref) -> bool:
    refs = ref_list(ref)
    if not refs:
        return False
    try:
        fi = page.locator('input[type="file"]').first
        fi.wait_for(state="attached", timeout=8000)
        fi.set_input_files(refs, timeout=8000)
        page.wait_for_timeout(4000 * len(refs))
        return True
    except Exception as e:
        print("   ref upload failed:", str(e).splitlines()[0])
        return False


def goto_chat(page, url):
    page.goto(url)
    page.wait_for_load_state("domcontentloaded")
    composer_loaded(page, 30000)
    page.wait_for_timeout(3000)
    try:
        page.evaluate("() => window.scrollTo(0, document.body.scrollHeight)")
    except Exception:
        pass
    page.wait_for_timeout(500)


def dismiss_dialog(page) -> bool:
    """ChatGPT occasionally throws a full-screen modal ("Compare responses" A/B eval)
    that overlays the composer and hangs the next prompt."""
    try:
        def is_open():
            return page.locator(
                'div[role="dialog"][data-state="open"], div[role="dialog"]').count() > 0
        if not is_open():
            return False
        print("   modal dialog present -> dismissing")
        for _ in range(3):
            page.keyboard.press("Escape")
            page.wait_for_timeout(700)
            if not is_open():
                return True
        for pat in (r"skip", r"no thanks", r"close", r"dismiss", r"done", r"prefer"):
            b = page.locator('div[role="dialog"] button',
                             has_text=re.compile(pat, re.IGNORECASE)).first
            if b.count():
                try:
                    b.click(timeout=4000)
                except Exception:
                    pass
                page.wait_for_timeout(700)
                if not is_open():
                    return True
        b0 = page.locator('div[role="dialog"] button').first
        if b0.count():
            try:
                b0.click(timeout=4000)
            except Exception:
                pass
            page.wait_for_timeout(700)
        if is_open():
            try:
                page.reload()
                page.wait_for_load_state("domcontentloaded")
            except Exception:
                pass
            page.wait_for_timeout(3000)
        return not is_open()
    except Exception:
        return False


class Generator:
    def __init__(self, prefix, images_base=None, registry=None, batch=None,
                 fake=False):
        self.prefix = prefix
        self.purpose = prefix                    # purpose == prefix, like the JS
        self.reg = registry                      # None = production registry
        self.batch = batch
        self.fake = fake
        base = Path(images_base) if images_base else IMG_BASE_DEFAULT
        self.outdir = base / subdir_for(prefix)
        self.outdir.mkdir(parents=True, exist_ok=True)
        self.page = None
        self.navigated_url = None    # the /c/ chat we're on (None = fresh chatgpt.com/)
        self.pending_fresh = False   # register its /c/ url after the first success
        self._route = f"**/*{IMAGE_URL_PATTERN}*"

    def out_path(self, item) -> Path:
        return self.outdir / f"{self.prefix}-{item['image_id']}-{item['slug']}.png"

    # ── chat management (pool-managed, port of ensureChat/openFresh) ──────────

    def open_fresh(self):
        page = self.page
        page.route(self._route, lambda r: r.abort())
        page.goto("https://chatgpt.com/")
        page.wait_for_load_state("domcontentloaded")
        composer_loaded(page, 30000)
        page.wait_for_timeout(2500)
        page.unroute(self._route)
        self.navigated_url, self.pending_fresh = None, True
        print("  opened a FRESH chat (will register its URL after first image)")

    def ensure_chat(self):
        active = pool.get_active_url(self.purpose, self.reg)
        if not active:
            if not (self.navigated_url is None and self.pending_fresh):
                self.open_fresh()
            return
        if active == self.navigated_url:
            return                                   # already on it, has room
        page = self.page
        page.route(self._route, lambda r: r.abort())  # block history images during load
        page.goto(active)
        page.wait_for_load_state("domcontentloaded")
        ok = composer_loaded(page, 12000)
        page.wait_for_timeout(2500)
        page.unroute(self._route)
        if not ok:
            print("  stored chat unreachable/deleted -> markDead + fresh")
            pool.mark_dead(self.purpose, self.reg)
            self.open_fresh()
            return
        self.navigated_url, self.pending_fresh = active, False
        print(f"  reusing pool chat ({pool.count_for(self.purpose, self.reg)}"
              f"/{pool.cap(self.reg)}): {active}")

    # ── one image (port of genOne, all gates intact) ──────────────────────────

    def gen_one(self, item):
        page = self.page
        out_path = self.out_path(item)
        if out_path.exists():
            print(f"SKIP (exists) {item['slug']}")
            return "skip"
        dismiss_dialog(page)
        composer = page.locator(COMPOSER_SEL).first
        composer.click()
        page.wait_for_timeout(600)
        # Upload the ref FIRST, THEN snapshot: the ref must be part of the `before`
        # baseline or it looks like a brand-new file_id and gets captured as the result.
        if item.get("ref"):
            upload_ref(page, item["ref"])
        before = {file_id(s) for s in get_gen_imgs(page)}
        composer.click()
        for ch in item["prompt"]:
            page.keyboard.type(ch)
            page.wait_for_timeout(random.randint(45, 69))
        page.wait_for_timeout(random.randint(6000, 9999))
        page.keyboard.press("Enter")
        # A fresh chat navigates chatgpt.com/ -> /c/<id> shortly AFTER the first send;
        # capture that url so a reload targets the generating chat.
        conv_url = page.url
        for _ in range(20):
            if "/c/" in conv_url:
                break
            page.wait_for_timeout(1000)
            conv_url = page.url

        # POST-SEND RE-BASELINE: the attachment is not in the DOM as an estuary <img>
        # until the message is POSTED, so it re-appears as "unseen" right after send.
        # Nothing real completes in ~3s; fold everything present now into `before`.
        page.wait_for_timeout(3000)
        for s in get_gen_imgs(page):
            before.add(file_id(s))

        def pick_new():
            imgs = get_gen_imgs(page)
            news = [s for s in imgs if file_id(s) not in before]
            est = [s for s in news if "estuary/content" in s]
            cand = est if est else news
            return cand[-1] if cand else None       # newest unseen generated image

        def finish(src) -> bool:
            buf = None
            try:
                r = page.request.get(src)
                if r.ok:
                    buf = r.body()
            except Exception:
                pass
            if not buf or len(buf) < 5000:
                return False
            # Never accept our own uploaded reference as the output: identical bytes
            # prove a mis-capture (a model cannot reproduce an input byte-for-byte).
            for rp in ref_list(item.get("ref")):
                rp = Path(rp)
                if not rp.exists():
                    continue
                rb = rp.read_bytes()
                if len(rb) == len(buf) and rb == buf:
                    print(f"   REJECT: captured the uploaded reference, not a render "
                          f"({item['slug']}) - still waiting")
                    return False
            for sib in self.outdir.glob("*.png"):
                sb = sib.read_bytes()
                if len(sb) == len(buf) and sb == buf:
                    print(f"FAIL (dup of {sib.name}) {item['slug']}")
                    return False
            out_path.write_bytes(buf)
            print(f"OK {item['slug']} ({len(buf) // 1024} KB)")
            return True

        # PHASE 1: poll the LIVE DOM up to 80s (fast path).
        t0 = time.monotonic()
        while time.monotonic() - t0 < 80:
            page.wait_for_timeout(5000)
            src = pick_new()
            if src:
                page.wait_for_timeout(3000)
                src = pick_new() or src
                if finish(src):
                    page.wait_for_timeout(2000)
                    return True
        # PHASE 2: hung past 80s -> RELOAD to surface the server-side-finished image.
        print(f"   >80s no new image in live DOM (hung) -> reloading to capture "
              f"{item['slug']}")
        reload_url = conv_url if "/c/" in conv_url else page.url
        t1 = time.monotonic()
        while time.monotonic() - t1 < 240:
            goto_chat(page, reload_url)
            dismiss_dialog(page)
            src = pick_new()
            if src and finish(src):
                page.wait_for_timeout(2000)
                return True
            page.wait_for_timeout(15000)
        print(f"FAIL (timeout) {item['slug']}")
        return False

    # ── fake mode (SANDBOX ONLY — the graph passes it under --test-sandbox) ───

    def gen_one_fake(self, item):
        out_path = self.out_path(item)
        if out_path.exists():
            print(f"SKIP (exists) {item['slug']}")
            return "skip"
        from PIL import Image, ImageDraw
        dims = {"ig-single": (1024, 1280)}.get(self.prefix, (1024, 1024))
        img = Image.new("RGB", dims, (10, 14, 30))
        dr = ImageDraw.Draw(img)
        dr.text((40, 40), f"FAKE SANDBOX {self.prefix}\n{item['image_id']}\n"
                          f"{item['slug']}", fill=(58, 244, 66))
        img.save(out_path)
        print(f"OK {item['slug']} (FAKE sandbox render)")
        return True

    # ── run a list ────────────────────────────────────────────────────────────

    def run(self, items) -> dict:
        ok = skip = fail = 0
        results = []
        if self.fake:
            print(f"gen_images: {len(items)} images | purpose=\"{self.purpose}\" | "
                  "** FAKE SANDBOX MODE — no browser, placeholder renders **")
            for i, item in enumerate(items):
                r = self.gen_one_fake(item)
                ok, skip = ok + (r is True), skip + (r == "skip")
                results.append((item, r))
                self._emit(item, r)
                print(f"PROGRESS {int((i + 1) * 100 / len(items))}%")
            print(f"GEN DONE ok={ok} skip={skip} fail={fail}")
            return {"ok": ok, "skip": skip, "fail": fail}

        from playwright.sync_api import sync_playwright
        print(f"gen_images: {len(items)} images | purpose=\"{self.purpose}\" | "
              f"cap={pool.cap(self.reg)} | current count="
              f"{pool.count_for(self.purpose, self.reg)}")
        with sync_playwright() as p:
            browser = p.chromium.launch_persistent_context(
                PROFILE_DIR, channel="chrome", headless=False,
                ignore_default_args=["--enable-automation"],
                args=["--disable-blink-features=AutomationControlled"],
                no_viewport=True)
            browser.add_init_script(
                "Object.defineProperty(navigator, 'webdriver', {get: () => undefined})")
            self.page = browser.new_page()
            try:
                for i, item in enumerate(items):
                    if self.out_path(item).exists():
                        print(f"SKIP (exists) {item['slug']}")
                        skip += 1
                        self._emit(item, "skip")
                        continue
                    self.ensure_chat()
                    r = self.gen_one(item)
                    if r is False:
                        print("   retry once...")   # side-effect-safe for image gen
                        r = self.gen_one(item)
                    if r is True:
                        if self.pending_fresh:
                            reg = pool.confirm_and_register(
                                self.page, self.purpose, self.batch, self.reg)
                            if reg:
                                self.navigated_url = reg["url"]
                                self.pending_fresh = False
                        pool.record_image(self.purpose, self.reg)
                        ok += 1
                    elif r == "skip":
                        skip += 1
                    else:
                        fail += 1
                    self._emit(item, r)
                    print(f"PROGRESS {int((i + 1) * 100 / len(items))}%")
                print(f"\nDone: {ok + skip}/{len(items)} | {self.purpose} chat now "
                      f"{pool.count_for(self.purpose, self.reg)}/{pool.cap(self.reg)}")
                # Delete rotated-out/dead chats while the browser is open. A sweep
                # failure never blocks the run — chats stay queued for the next sweep.
                try:
                    chat_delete.sweep_retired(self.page, self.reg)
                except Exception as e:
                    print("  [chat-delete] sweep error: " + str(e).splitlines()[0])
            finally:
                browser.close()
        print(f"GEN DONE ok={ok} skip={skip} fail={fail}")
        return {"ok": ok, "skip": skip, "fail": fail}

    def _emit(self, item, r):
        out = self.out_path(item)
        if r is True:
            print(f"IMG OK purpose={self.prefix} id={item['image_id']} "
                  f"slug={item['slug']} bytes={out.stat().st_size} out={out}")
        elif r == "skip":
            print(f"IMG SKIP purpose={self.prefix} id={item['image_id']} "
                  f"slug={item['slug']} (exists)")
        else:
            print(f"IMG FAIL purpose={self.prefix} id={item['image_id']} "
                  f"slug={item['slug']} reason=no-capture")


def main():
    ap = argparse.ArgumentParser(
        description="Pool-managed ChatGPT image generator (canonical Python port).")
    ap.add_argument("--list", required=True, help="items JSON path")
    ap.add_argument("--prefix", default="x-tweets",
                    choices=["x-tweets", "yt-posts", "ig-single", "ig-carousel"])
    ap.add_argument("--images-base", default=None,
                    help="override schedule-tweets/images (sandbox)")
    ap.add_argument("--registry", default=None,
                    help="override the chat registry path (sandbox)")
    ap.add_argument("--batch", default=None,
                    help="batches.json id to tie a fresh chat to (cleanup deletes it "
                         "when the batch completes)")
    ap.add_argument("--fake", action="store_true",
                    help="SANDBOX ONLY: placeholder renders, no browser")
    args = ap.parse_args()

    items = json.loads(Path(args.list).read_text(encoding="utf-8"))
    g = Generator(args.prefix, images_base=args.images_base, registry=args.registry,
                  batch=args.batch, fake=args.fake)
    res = g.run(items)
    sys.exit(1 if res["fail"] else 0)


if __name__ == "__main__":
    main()
